### ISSUE TITLE
Error using numba-scipy extension to calculate cdist in python

### ISSUE THREAD

**Issue by mrgransky** [numba/numba-scipy#38]
I'd like to speed up the `cdist` between two `numpy.ndarray` using `numba` as follows:

    import numpy as np
    from numba import njit, jit
    from scipy.spatial.distance import cdist
    import time
        
    @njit
    def dist_scipy(a, b):
    	d = cdist(a, b, 'euclidean')
    	d = np.transpose(d)
    	sorted_d = np.sort(d)
    	sorted_ind = np.argsort(d)
    	return sorted_d, sorted_ind
    
    def get_a_b(r=10**4,c=10** 1):
    	a = np.random.uniform(-1, 1, (r, c)).astype('f')
    	b = np.random.uniform(-1, 1, (r, c)).astype('f')
    	return a,b
    
    if __name__ == "__main__":
        a, b = get_a_b()
        st_t = time.time()
        dist_scipy(a,b)
        print('it took {} s'.format(time.time()-st_t))
After `$ pip install numba-scipy`, I get the following error:

    Traceback (most recent call last):
      File "stackoverflow_Q.py", line 31, in <module>
        dist_scipy(a,b)
      File "/usr/local/lib/python2.7/dist-packages/numba/dispatcher.py", line 420, in _compile_for_args
        raise e
      File "/usr/local/lib/python2.7/dist-packages/numba_scipy/special/overloads.py", line 12
        f = signatures.name_and_types_to_pointer[(name, *signature)]
                                                        ^
    SyntaxError: invalid syntax

**Comment by stuartarchibald** [comment]
Thanks for the report. I think the error you are seeing is because you are using Python 2, which was [retired on 1st January 2020](https://www.python.org/doc/sunset-python-2/), and some Python 3 syntax in `numba-scipy` is not compatible with Python 2. As to the feature request, thanks, there's a few people asking for `scipy.spatial` functions, will mark it this as a request.

**Comment by mrgransky** [comment]
Error with conda environment and python3:

```
Traceback (most recent call last):
  File "numba_scipy_test.py", line 31, in <module>
    dist_scipy(a,b)
  File "/home/alijani/.conda/envs/py3_gpu/lib/python3.7/site-packages/numba/dispatcher.py", line 420, in _compile_for_args
    raise e
  File "/home/alijani/.conda/envs/py3_gpu/lib/python3.7/site-packages/numba/dispatcher.py", line 353, in _compile_for_args
    return self.compile(tuple(argtypes))
  File "/home/alijani/.conda/envs/py3_gpu/lib/python3.7/site-packages/numba/compiler_lock.py", line 32, in _acquire_compile_lock
    return func(*args, **kwargs)
  File "/home/alijani/.conda/envs/py3_gpu/lib/python3.7/site-packages/numba/dispatcher.py", line 768, in compile
    cres = self._compiler.compile(args, return_type)
  File "/home/alijani/.conda/envs/py3_gpu/lib/python3.7/site-packages/numba/dispatcher.py", line 77, in compile
    status, retval = self._compile_cached(args, return_type)
  File "/home/alijani/.conda/envs/py3_gpu/lib/python3.7/site-packages/numba/dispatcher.py", line 91, in _compile_cached
    retval = self._compile_core(args, return_type)
  File "/home/alijani/.conda/envs/py3_gpu/lib/python3.7/site-packages/numba/dispatcher.py", line 109, in _compile_core
    pipeline_class=self.pipeline_class)
  File "/home/alijani/.conda/envs/py3_gpu/lib/python3.7/site-packages/numba/compiler.py", line 550, in compile_extra
    args, return_type, flags, locals)
  File "/home/alijani/.conda/envs/py3_gpu/lib/python3.7/site-packages/numba/compiler.py", line 281, in __init__
    targetctx.refresh()
  File "/home/alijani/.conda/envs/py3_gpu/lib/python3.7/site-packages/numba/targets/base.py", line 281, in refresh
    self.load_additional_registries()
  File "/home/alijani/.conda/envs/py3_gpu/lib/python3.7/site-packages/numba/targets/cpu.py", line 80, in load_additional_registries
    numba.entrypoints.init_all()
  File "/home/alijani/.conda/envs/py3_gpu/lib/python3.7/site-packages/numba/entrypoints.py", line 24, in init_all
    func()
  File "/home/alijani/.conda/envs/py3_gpu/lib/python3.7/site-packages/numba_scipy/__init__.py", line 12, in _init_extension
    from . import special
  File "/home/alijani/.conda/envs/py3_gpu/lib/python3.7/site-packages/numba_scipy/special/__init__.py", line 1, in <module>
    from . import overloads as _overloads
  File "/home/alijani/.conda/envs/py3_gpu/lib/python3.7/site-packages/numba_scipy/special/overloads.py", line 4, in <module>
    from . import signatures
  File "/home/alijani/.conda/envs/py3_gpu/lib/python3.7/site-packages/numba_scipy/special/signatures.py", line 376, in <module>
    ('pdtr', numba.types.float64, numba.types.float64): ctypes.CFUNCTYPE(ctypes.c_double, ctypes.c_double, ctypes.c_double)(get_cython_function_address('scipy.special.cython_special', '__pyx_fuse_0pdtr')),
  File "/home/alijani/.conda/envs/py3_gpu/lib/python3.7/site-packages/numba/extending.py", line 406, in get_cython_function_address
    return _import_cython_function(module_name, function_name)
ValueError: No function '__pyx_fuse_0pdtr' found in __pyx_capi__ of 'scipy.special.cython_special'
```

**Comment by max9111** [comment]
It is quite straight forward to implement cdist(metric="euclidean"). Three possible ways to implement this:

```Python
import numba as nb
import numpy as np
from scipy.spatial.distance import cdist

@nb.njit(fastmath=True,parallel=True)
def eucl_naive(A,B):
    assert A.shape[1]==B.shape[1]
    C=np.empty((A.shape[0],B.shape[0]),A.dtype)
    
    #workaround to get the right datatype for acc
    init_val_arr=np.zeros(1,A.dtype)
    init_val=init_val_arr[0]
    
    for i in nb.prange(A.shape[0]):
        for j in range(B.shape[0]):
            acc=init_val
            for k in range(A.shape[1]):
                acc+=(A[i,k]-B[j,k])**2
            C[i,j]=np.sqrt(acc)
    return C

@nb.njit(fastmath=True,parallel=True)
def eucl_opt(A, B):
    assert A.shape[1]==B.shape[1]
    C=np.empty((A.shape[0],B.shape[0]),A.dtype)
    I_BLK=32
    J_BLK=32
    
    #workaround to get the right datatype for acc
    init_val_arr=np.zeros(1,A.dtype)
    init_val=init_val_arr[0]
    
    #Blocking and partial unrolling
    #Beneficial if the second dimension is large -> computationally bound problem 
    # 
    for ii in nb.prange(A.shape[0]//I_BLK):
        for jj in range(B.shape[0]//J_BLK):
            for i in range(I_BLK//4):
                for j in range(J_BLK//2):
                    acc_0=init_val
                    acc_1=init_val
                    acc_2=init_val
                    acc_3=init_val
                    acc_4=init_val
                    acc_5=init_val
                    acc_6=init_val
                    acc_7=init_val
                    for k in range(A.shape[1]):
                        acc_0+=(A[ii*I_BLK+i*4+0,k] - B[jj*J_BLK+j*2+0,k])**2
                        acc_1+=(A[ii*I_BLK+i*4+0,k] - B[jj*J_BLK+j*2+1,k])**2
                        acc_2+=(A[ii*I_BLK+i*4+1,k] - B[jj*J_BLK+j*2+0,k])**2
                        acc_3+=(A[ii*I_BLK+i*4+1,k] - B[jj*J_BLK+j*2+1,k])**2
                        acc_4+=(A[ii*I_BLK+i*4+2,k] - B[jj*J_BLK+j*2+0,k])**2
                        acc_5+=(A[ii*I_BLK+i*4+2,k] - B[jj*J_BLK+j*2+1,k])**2
                        acc_6+=(A[ii*I_BLK+i*4+3,k] - B[jj*J_BLK+j*2+0,k])**2
                        acc_7+=(A[ii*I_BLK+i*4+3,k] - B[jj*J_BLK+j*2+1,k])**2
                    C[ii*I_BLK+i*4+0,jj*J_BLK+j*2+0]=np.sqrt(acc_0)
                    C[ii*I_BLK+i*4+0,jj*J_BLK+j*2+1]=np.sqrt(acc_1)
                    C[ii*I_BLK+i*4+1,jj*J_BLK+j*2+0]=np.sqrt(acc_2)
                    C[ii*I_BLK+i*4+1,jj*J_BLK+j*2+1]=np.sqrt(acc_3)
                    C[ii*I_BLK+i*4+2,jj*J_BLK+j*2+0]=np.sqrt(acc_4)
                    C[ii*I_BLK+i*4+2,jj*J_BLK+j*2+1]=np.sqrt(acc_5)
                    C[ii*I_BLK+i*4+3,jj*J_BLK+j*2+0]=np.sqrt(acc_6)
                    C[ii*I_BLK+i*4+3,jj*J_BLK+j*2+1]=np.sqrt(acc_7)
        #Remainder j
        for i in range(I_BLK):
            for j in range((B.shape[0]//J_BLK)*J_BLK,B.shape[0]):
                acc_0=init_val
                for k in range(A.shape[1]):
                    acc_0+=(A[ii*I_BLK+i,k] - B[j,k])**2
                C[ii*I_BLK+i,j]=np.sqrt(acc_0)
    
    #Remainder i
    for i in range((A.shape[0]//I_BLK)*I_BLK,A.shape[0]):
        for j in range(B.shape[0]):
            acc_0=init_val
            for k in range(A.shape[1]):
                acc_0+=(A[i,k] - B[j,k])**2
            C[i,j]=np.sqrt(acc_0)
            
    return C

@nb.njit(fastmath=True)
def eucl_BLAS(A,B):
    dist=np.dot(A,B.T)

    TMP_A=np.empty(A.shape[0],dtype=A.dtype)
    for i in range(A.shape[0]):
        sum=0
        for j in range(A.shape[1]):
            sum+=A[i,j]**2
        TMP_A[i]=sum

    TMP_B=np.empty(B.shape[0],dtype=A.dtype)
    for i in range(B.shape[0]):
        sum=0
        for j in range(B.shape[1]):
            sum+=B[i,j]**2
        TMP_B[i]=sum
    
    for i in range(A.shape[0]):
        for j in range(B.shape[0]):
            dist[i,j]=np.sqrt(-2.*dist[i,j]+TMP_A[i]+TMP_B[j])
    return dist
```

**Timings**

```Python
A=np.random.rand(10_000,10).astype("f")
B=np.random.rand(10_000,10).astype("f")

%timeit cdist(A,B,metric="euclidean")
#1.09 s ± 4.23 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
%timeit eucl_naive(A,B)
#146 ms ± 695 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
%timeit eucl_opt(A,B)
#105 ms ± 5.15 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
print(np.allclose(eucl_opt(A,B),cdist(A,B,metric="euclidean")))
#True
#Using this method higher numerical errors could occur
%timeit eucl_approx(A,B)
#186 ms ± 580 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

#larger example
A=np.random.rand(10_000,1000).astype("f")
B=np.random.rand(10_000,1000).astype("f")

%timeit cdist(A,B,metric="euclidean")
#1min 43s ± 338 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
%timeit eucl_naive(A,B)
#4.36 s ± 279 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
%timeit eucl_opt(A,B)
#988 ms ± 10.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
print(np.allclose(eucl_opt(A,B),cdist(A,B,metric="euclidean")))
#True

#Using this method higher numerical errors could occur
%timeit eucl_approx(A,B)
#501 ms ± 11.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

```

**Comment by mrgransky** [comment]
@max9111 did u try your code with `python3.X` ? cuz I am getting this error using `python2.X` : 
```
Traceback (most recent call last):
  File "distance_error.py", line 129, in <module>
    eucl_opt(rf, qu)
  File "/usr/local/lib/python2.7/dist-packages/numba/dispatcher.py", line 420, in _compile_for_args
    raise e
  File "/usr/local/lib/python2.7/dist-packages/numba_scipy/special/overloads.py", line 12
    f = signatures.name_and_types_to_pointer[(name, *signature)]
                                                    ^
SyntaxError: invalid syntax
```

**Comment by esc** [comment]
The last error on this `ValueError: No function '__pyx_fuse_0pdtr' found in __pyx_capi__ of 'scipy.special.cython_special'` has most likely been resolved with #54 -- thus I am closing this issue now. But feel free to re-open in case I am mistaken and if the problem still exists. Thank you!
